### PR TITLE
Some ingest cleanups

### PR DESCRIPTION
**Pull request by martijnvg**
- Inlined PipelineStoreClient class into the PipelineStore class
- Moved PipelineReference to a top level class and named it PipelineDefinition
- Pulled some logic from the crud transport classes to the PipelineStore
- Use IOUtils#close(...) where appropriate

Based on the review #15007 and https://github.com/s1monw/elasticsearch/commit/042ef8c20476459f92ea1c2ea12bc1950211e693

### CODE REVIEW

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineDefinition.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.ingest.Pipeline;
+
+import java.io.IOException;
+
+public class PipelineDefinition implements Streamable, ToXContent {
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineDefinition.java]: use Writeable and call it WriteablePipeline ?

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java [hunk 2]
```diff
@@ -69,6 +91,7 @@ public PipelineStore(Settings settings, ThreadPool threadPool, Environment envir
 
     @Override
     protected void doStart() {
+        client = injector.getInstance(Client.class);
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java]: wonder if this is the best way to get a client, can't come up with better ones though at the moment....

Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java]: Simon suggested that we may be able to get it as a constructor argument? or would that create a cyclic dependency? maybe not if we get rid of the ClusterService dependency?

Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java]: I wish there was a better way... if the `Client` gets injected either via setter or constructor then there is an cyclic dependency error. 

(the node client can't be created until all transport services (transport actions, action filters etc.) have been created. The IngestActionFilter relies indirectly on PipelineStore which has Client as a dependency. )

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java [hunk 3]
```diff
@@ -165,6 +214,75 @@ void startUpdateWorker() {
...
+        SearchResponse searchResponse = client.search(searchRequest).actionGet(scrollTimeout);
+
+        if (searchResponse.getHits().getTotalHits() == 0) {
+            return Collections.emptyList();
+        }
+        logger.debug("reading [{}] pipeline documents", searchResponse.getHits().totalHits());
+        return () -> new SearchScrollIterator(searchResponse);
+    }
+
+    class SearchScrollIterator implements Iterator<SearchHit> {
```
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java]: would it make sense to add this nice iterator to core? the ingest plugin could reuse it in this case.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java (hunk 2 above)
Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java]: ok thanks for the explanation. @s1monw any magic that we can do to fix this? :)

**Comment by javanna** [comment]
I like it, left some comments but looks good

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java (hunk 3 above)
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java]: yes, wanted to do this in a followup PR, but can it do now too.

File: plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineDefinition.java (hunk 1 above)
Comment (martijnvg) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineDefinition.java]: this get serialised on the transport layer so it should implement `Streamable`, right?

Comment (javanna) [inline plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineDefinition.java]: Writeable does that too but allows to have final fields and drop empty constructors

**Comment by martijnvg** [comment]
@javanna I've updated the PR and make use of Writable instead of Streamable and moved the scroll search iterator to core as helper class.

File: plugins/ingest/src/test/java/org/elasticsearch/plugin/ingest/PipelineStoreTests.java [hunk 4]
```diff
@@ -173,4 +184,49 @@ public int compare(PipelineStore.PipelineReference first, PipelineStore.Pipeline
...
+                return new GetResponse(new GetResult("_index", "_type", "_id", 1, exists, null, null));
+            }
+        };
+    }
+
+    GetRequest eqGetRequest(String index, String type, String id) {
+        return Matchers.argThat(new GetRequestMatcher(index, type, id));
+    }
+
+    static class GetRequestMatcher extends ArgumentMatcher<GetRequest> {
```
Comment (javanna) [inline plugins/ingest/src/test/java/org/elasticsearch/plugin/ingest/PipelineStoreTests.java]: do we really need this matcher to check one get request? maybe we can instead do it inline or something?

**Comment by javanna** [comment]
left a super small comment on testing, LGTM otherwise

File: plugins/ingest/src/test/java/org/elasticsearch/plugin/ingest/PipelineStoreTests.java (hunk 4 above)
Comment (martijnvg) [inline plugins/ingest/src/test/java/org/elasticsearch/plugin/ingest/PipelineStoreTests.java]: I can make it an anonymous inner class, but then the method gets less clean IMO, so I prefer it to be a separate inner class.

**Pull request merged**